winch/codegen/src/codegen/context.rs

```diff
@@ -46,24 +48,63 @@ impl<'a> CodeGenContext<'a> {
         }
     }
 
-    /// Request a specific general purpose register to the register allocator,
+    /// Request a specific register to the register allocator,
     /// spilling if not available.
-    pub fn gpr<M: MacroAssembler>(&mut self, named: Reg, masm: &mut M) -> Reg {
-        self.regalloc.gpr(named, &mut |regalloc| {
+    pub fn reg<M: MacroAssembler>(&mut self, named: Reg, masm: &mut M) -> Reg {
+        self.regalloc.reg(named, |regalloc| {
             Self::spill_impl(&mut self.stack, regalloc, &self.frame, masm)
         })
     }
 
+    /// Allocate a register for the given WebAssembly type.
+    pub fn reg_for_type<M: MacroAssembler>(&mut self, ty: WasmType, masm: &mut M) -> Reg {
+        use WasmType::*;
+        match ty {
+            I32 | I64 => self.any_gpr(masm),
+            F32 | F64 => self.any_fpr(masm),
+            t => panic!("unsupported type {:?}", t),
+        }
+    }
+
     /// Request the next avaiable general purpose register to the register allocator,
     /// spilling if no registers are available.
     pub fn any_gpr<M: MacroAssembler>(&mut self, masm: &mut M) -> Reg {
         self.regalloc
             .any_gpr(&mut |regalloc| Self::spill_impl(&mut self.stack, regalloc, &self.frame, masm))
     }
 
-    /// Free the given general purpose register.
-    pub fn free_gpr(&mut self, reg: Reg) {
-        self.regalloc.free_gpr(reg);
+    /// Request the next avaiable floating point register to the register allocator,
+    /// spilling if no registers are available.
+    pub fn any_fpr<M: MacroAssembler>(&mut self, masm: &mut M) -> Reg {
```

Comment:
rather than duplicate code-paths for `gpr` and `fpr` variants, would it make sense to have an array indexed by class? Either regalloc2's `RegClass` could be used or a custom type here, doesn't matter too much...

Reply:
I've added class based indexing to `RegSet`, it indeed removed quite some duplication. I've opted to keep `CodeGenContext::any_gpr` and rewrote it in terms of `CodeGenContext::reg_for_class`